sdk/core/Azure.Core.Experimental/tests/Auth/TokenProviderTests.cs

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Auth;
+using System.ClientModel.Primitives;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.TestFramework;
+using ClientModel.Tests.Mocks;
+using NUnit.Framework;
+
+namespace System.ClientModel.Tests.Auth;
+
+public class TokenProviderTests
+{
+    [Test]
+    public void SampleUsage()
+    {
+        // usage for TokenProvider2 abstract type
+        ITokenProvider provider = new ClientCredentialTokenProvider();
+        var client = new FooClient(new Uri("http://localhost"), provider);
+        client.Get();
+    }
+
+    public class FooClient
+    {
+        // Generated from the TypeSpec spec.
+        private readonly Dictionary<string, object>[] flows = [
+            new Dictionary<string, object> {
+                { "scopes", new string[] { "baselineScope" } },
+                { "tokenUrl" , new Uri("https://myAuthserver.com/token")},
+                { "refreshUrl" , new Uri("https://myAuthserver.com/refresh")}
+            }
+        ];
+
+        private ClientPipeline _pipeline;
+
+        public FooClient(Uri uri, ApiKeyCredential credential)
+        {
+            var options = new ClientPipelineOptions();
+            options.Transport = new MockPipelineTransport("foo", m => new MockPipelineResponse(200));
+            ClientPipeline pipeline = ClientPipeline.Create(options,
+            perCallPolicies: ReadOnlySpan<PipelinePolicy>.Empty,
+            perTryPolicies: [ApiKeyAuthenticationPolicy.CreateBasicAuthorizationPolicy(credential)],
+            beforeTransportPolicies: ReadOnlySpan<PipelinePolicy>.Empty);
+            _pipeline = pipeline;
+        }
+
+        public FooClient(Uri uri, ITokenProvider credential)
+        {
+            var options = new ClientPipelineOptions();
+            options.Transport = new MockPipelineTransport("foo",
+            m =>
+            {
+                return new MockPipelineResponse(200);
+            });
+            ClientPipeline pipeline = ClientPipeline.Create(options,
+            perCallPolicies: ReadOnlySpan<PipelinePolicy>.Empty,
+            perTryPolicies: [new OAuth2BearerTokenAuthenticationPolicy(credential, flows)],
+            beforeTransportPolicies: ReadOnlySpan<PipelinePolicy>.Empty);
+            _pipeline = pipeline;
+        }
+
+        public ClientResult Get()
+        {
+            var message = _pipeline.CreateMessage();
+            message.ResponseClassifier = PipelineMessageClassifier.Create(stackalloc ushort[] { 200 });
+            message.SetProperty(typeof(IScopedFlowToken), new ScopedContext(["read"]));
```

Comment:
example of operation specific scope being attached to the message. These scopes will be applied to the token provider context by the auth policy.